### ISSUE TITLE
SSL3 alert write:fatal:bad record mac by RC4-SHA

### ISSUE THREAD

**Issue by 1185129439** [openssl/openssl#13363]
`server:
OPENSSL_MODULES=../providers gdb --args   ./openssl s_server -tls1 -state  -cert eecert.pem -key eekey.pem   -cipher RC4-SHA -msg`

`client :
OPENSSL_MODULES=../providers gdb --args  ./openssl s_client -tls1 -state -cipher RC4-SHA -CAfile rootcert.pem -CAfile cacert.pem -msg`

I loaded the legacy provider into the default library context as shown below
`       legacy = OSSL_PROVIDER_load(NULL, "legacy");
       if (legacy == NULL) {
           printf("Failed to load Legacy provider\n");
       }
       deflt = OSSL_PROVIDER_load(NULL, "default");
       if (deflt == NULL) {
           printf("Failed to load Default provider\n");
           OSSL_PROVIDER_unload(legacy);
       }
   `

And set security level 0 SSL_CTX_set_security_level(ctx, 0)

But there  appeared the following error
`client:
SSL3 alert read:fatal:bad record mac
SSL_connect:error in error
0017FDF7FF7F0000:error::SSL routines::sslv3 alert bad record mac:ssl/record/rec_layer_s3.c:1615:SSL alert number 20`

`server:
SSL3 alert write:fatal:bad record mac
SSL_accept:error in error
ERROR
0017FDF7FF7F0000:error::SSL routines::decryption failed or bad record mac:ssl/record/ssl3_record.c:664:`

Can you give me a help,thanks!

**Comment by 1185129439** [comment]
I'm using alpha7

**Comment by paulidale** [comment]
You need to load the legacy provider.  Add `-provider legacy` to the command line.

**Comment by 1185129439** [comment]
The client log：
`>>> TLS 1.0, RecordHeader [length 0005]
    16 03 01 00 4f
>>> TLS 1.0, Handshake [length 004f], ClientHello
    01 00 00 4b 03 01 22 c4 70 e3 4f 46 d8 91 d5 58
    fb f0 e2 05 e4 d7 21 42 dd 2a 37 77 f4 d7 00 74
    3b a8 37 f9 45 68 00 00 04 00 05 00 ff 01 00 00
    1e 00 00 00 0e 00 0c 00 00 09 6c 6f 63 61 6c 68
    6f 73 74 00 23 00 00 00 16 00 00 00 17 00 00
SSL_connect:SSLv3/TLS write client hello
<<< TLS 1.0, RecordHeader [length 0005]
    16 03 01 00 39
SSL_connect:SSLv3/TLS write client hello
<<< TLS 1.0, Handshake [length 0039], ServerHello
    02 00 00 35 03 01 fa e5 c4 52 94 33 1c 53 fd ef
    1d b0 e6 6d 75 4c 52 26 c8 a6 0f d5 24 4d 93 7e
    d1 a7 0a 07 1d f1 00 00 05 00 00 0d ff 01 00 01
    00 00 23 00 00 00 17 00 00
<<< TLS 1.0, RecordHeader [length 0005]
    16 03 01 04 01
SSL_connect:SSLv3/TLS read server hello
<<< TLS 1.0, Handshake [length 0401], Certificate
    0b 00 03 fd 00 03 fa 00 03 f7 30 82 03 f3 30 82
    03 5c a0 03 02 01 02 02 09 00 94 9e 8b c5 d7 ff
    d8 c7 30 0d 06 09 2a 86 48 86 f7 0d 01 01 05 05
    00 30 79 31 0b 30 09 06 03 55 04 06 13 02 69 6e
    31 0c 30 0a 06 03 55 04 08 13 03 6b 61 72 31 0c
    30 0a 06 03 55 04 07 13 03 62 61 6e 31 0f 30 0d
    06 03 55 04 0a 13 06 68 75 61 77 65 69 31 0d 30
    0b 06 03 55 04 0b 13 04 69 70 73 69 31 0b 30 09
    06 03 55 04 03 13 02 63 61 31 21 30 1f 06 09 2a
    86 48 86 f7 0d 01 09 01 16 12 6f 70 65 6e 73 73
    6c 40 68 75 61 77 65 69 2e 63 6f 6d 30 1e 17 0d
    30 39 30 34 32 34 31 31 35 35 34 30 5a 17 0d 32
    39 30 34 31 39 31 31 35 35 34 30 5a 30 79 31 0b
    30 09 06 03 55 04 06 13 02 69 6e 31 0c 30 0a 06
    03 55 04 08 13 03 6b 61 72 31 0c 30 0a 06 03 55
    04 07 13 03 62 61 6e 31 0f 30 0d 06 03 55 04 0a
    13 06 68 75 61 77 65 69 31 0d 30 0b 06 03 55 04
    0b 13 04 69 70 73 69 31 0b 30 09 06 03 55 04 03
    13 02 65 65 31 21 30 1f 06 09 2a 86 48 86 f7 0d
    01 09 01 16 12 6f 70 65 6e 73 73 6c 40 68 75 61
    77 65 69 2e 63 6f 6d 30 81 9f 30 0d 06 09 2a 86
    48 86 f7 0d 01 01 01 05 00 03 81 8d 00 30 81 89
    02 81 81 00 fa ef b5 44 15 0f 1b 7e d2 26 1c 90
    5a 29 43 a0 0f 5b 2a 90 5e d8 14 ec 96 53 1d 41
    13 e6 6a f4 ba ee d0 b2 74 71 83 71 d5 7f 9f db
    40 a9 7e b0 74 e7 6a 06 58 c7 e0 ee 60 96 1f 55
    73 06 0f 22 6a 71 56 35 70 6a 3f 2f 15 73 9e b5
    67 46 c3 eb d8 1f b9 41 43 73 9b 3e 57 e5 e2 f6
    60 04 4d 0a 7c ee d6 46 a2 cc a1 82 d0 f2 43 b0
    f0 ac 1c da ab 42 5e 80 8e ba 2f a0 bd 49 53 9c
    9e 75 37 f7 02 03 01 00 01 a3 82 01 81 30 82 01
    7d 30 09 06 03 55 1d 13 04 02 30 00 30 0e 06 03
    55 1d 0f 01 01 ff 04 04 03 02 05 e0 30 42 06 03
    55 1d 1f 04 3b 30 39 30 1a a0 18 a0 16 86 14 68
    74 74 70 3a 2f 2f 6d 79 2e 63 6f 6d 2f 6d 79 2e
    63 72 6c 30 1b a0 19 a0 17 86 15 68 74 74 70 3a
    2f 2f 6f 74 68 2e 63 6f 6d 2f 6d 79 2e 63 72 6c
    30 2c 06 09 60 86 48 01 86 f8 42 01 0d 04 1f 16
    1d 4f 70 65 6e 53 53 4c 20 47 65 6e 65 72 61 74
    65 64 20 43 65 72 74 69 66 69 63 61 74 65 30 1d
    06 03 55 1d 0e 04 16 04 14 63 fc eb fd d8 2d 99
    05 ec d2 0c 54 6f 3f a6 5c b3 f4 6f 2d 30 81 ad
    06 03 55 1d 23 04 81 a5 30 81 a2 80 14 02 dd 2e
    2a aa 13 2b 3b 2d 5a b2 a5 7b 09 86 e3 c6 dd 7f
    a8 a1 7f a4 7d 30 7b 31 0b 30 09 06 03 55 04 06
    13 02 69 6e 31 0c 30 0a 06 03 55 04 08 13 03 6b
    61 72 31 0c 30 0a 06 03 55 04 07 13 03 62 61 6e
    31 0f 30 0d 06 03 55 04 0a 13 06 68 75 61 77 65
    69 31 0d 30 0b 06 03 55 04 0b 13 04 69 70 73 69
    31 0d 30 0b 06 03 55 04 03 13 04 72 6f 6f 74 31
    21 30 1f 06 09 2a 86 48 86 f7 0d 01 09 01 16 12
    6f 70 65 6e 73 73 6c 40 68 75 61 77 65 69 2e 63
    6f 6d 82 09 00 ac 4e 0f e8 f4 fa 69 91 30 1f 06
    03 55 1d 11 04 18 30 16 87 04 c0 a8 07 01 82 0e
    77 77 77 2e 68 75 61 77 65 69 2e 63 6f 6d 30 0d
    06 09 2a 86 48 86 f7 0d 01 01 05 05 00 03 81 81
    00 2a 68 a2 11 10 a8 92 98 54 7e 1e cc 58 29 68
    88 3f 4a 15 d1 71 24 44 f2 0b 87 09 f2 59 3b e3
    c3 57 e4 ce 33 08 2a f7 66 b0 0e 73 dc f7 01 93
    6c 4d 92 a4 f6 91 ac fe fc 32 93 33 56 c0 a1 55
    36 a5 5d d2 c8 45 32 97 17 70 8b 80 59 a6 74 51
    37 53 d6 31 fa 6e 2f 11 9b 12 8e a0 04 90 8c c0
    f9 23 5c 3c fc 16 5c 93 40 19 28 4e ad 22 ca 5f
    c1 7e 29 fa 33 60 8b e7 60 9a f8 48 54 34 a5 0c
    a7
depth=1 C = in, ST = kar, L = ban, O = huawei, OU = ipsi, CN = ca, emailAddress = openssl@xx.com
verify error:num=2:unable to get issuer certificate
issuer= C = in, ST = kar, L = ban, O = huawei, OU = ipsi, CN = root, emailAddress = openssl@xx.com
verify return:1
depth=0 C = in, ST = kar, L = ban, O = huawei, OU = ipsi, CN = ee, emailAddress = openssl@xx.com
verify error:num=2:unable to get issuer certificate
issuer= C = in, ST = kar, L = ban, O = huawei, OU = ipsi, CN = ca, emailAddress = openssl@xx.com
verify return:1
depth=1 C = in, ST = kar, L = ban, O = huawei, OU = ipsi, CN = ca, emailAddress = openssl@xx.com
verify error:num=2:unable to get issuer certificate
issuer= C = in, ST = kar, L = ban, O = huawei, OU = ipsi, CN = root, emailAddress = openssl@xx.com
verify return:1
depth=0 C = in, ST = kar, L = ban, O = huawei, OU = ipsi, CN = ee, emailAddress = openssl@xx.com
issuer= C = in, ST = kar, L = ban, O = huawei, OU = ipsi, CN = ca, emailAddress = openssl@xx.com
verify return:1
<<< TLS 1.0, RecordHeader [length 0005]
    16 03 01 00 04
SSL_connect:SSLv3/TLS read server certificate
<<< TLS 1.0, Handshake [length 0004], ServerHelloDone
    0e 00 00 00
SSL_connect:SSLv3/TLS read server done
>>> TLS 1.0, RecordHeader [length 0005]
    16 03 01 00 86
>>> TLS 1.0, Handshake [length 0086], ClientKeyExchange
    10 00 00 82 00 80 c7 04 b5 4e bf 34 18 6b 32 09
    77 51 41 ac 39 d5 03 92 56 3a 85 17 d1 dd 3e 85
    92 77 84 22 fa c4 cd 5d 79 67 79 95 6f 3f 2e de
    fe d3 64 63 e0 57 8c ac c5 57 fd 4a 1d de 01 81
    ec ca f3 63 de 8c d3 c9 64 c4 61 e1 80 dd 16 f3
    61 ef b0 c3 cb ac e5 11 3c 88 15 0e 66 1a 27 87
    49 69 63 bc cc df e7 19 53 a9 4a 4e f0 1b 69 55
    6d 66 63 13 42 91 64 49 28 cf d3 39 19 84 ed d6
    38 13 6b e6 a7 77
SSL_connect:SSLv3/TLS write client key exchange
>>> TLS 1.0, RecordHeader [length 0005]
    14 03 01 00 01
>>> TLS 1.0, ChangeCipherSpec [length 0001]
    01
SSL_connect:SSLv3/TLS write change cipher spec
>>> TLS 1.0, RecordHeader [length 0005]
    16 03 01 00 24
>>> TLS 1.0, Handshake [length 0010], Finished
    14 00 00 0c 3f 11 59 03 7a 20 a7 67 17 3e ec db
SSL_connect:SSLv3/TLS write finished
<<< TLS 1.0, RecordHeader [length 0005]
    15 03 01 00 02
<<< TLS 1.0, Alert [length 0002], fatal bad_record_mac
    02 14
SSL3 alert read:fatal:bad record mac
SSL_connect:error in error
20F46EBFFFFF0000:error::SSL routines::sslv3 alert bad record mac:ssl/record/rec_layer_s3.c:1615:SSL alert number 20
---
Certificate chain
 0 s:C = in, ST = kar, L = ban, O = huawei, OU = ipsi, CN = ee, emailAddress = openssl@xx.com
   i:C = in, ST = kar, L = ban, O = huawei, OU = ipsi, CN = ca, emailAddress = openssl@xx.com
   a:PKEY: rsaEncryption, 1024 (bit); sigalg: RSA-SHA1
   v:NotBefore: Apr 24 11:55:40 2009 GMT; NotAfter: Apr 19 11:55:40 2029 GMT
---
Server certificate
-----BEGIN CERTIFICATE-----
MIID8zCCA1ygAwIBAgIJAJSei8XX/9jHMA0GCSqGSIb3DQEBBQUAMHkxCzAJBgNV
BAYTAmluMQwwCgYDVQQIEwNrYXIxDDAKBgNVBAcTA2JhbjEPMA0GA1UEChMGaHVh
d2VpMQ0wCwYDVQQLEwRpcHNpMQswCQYDVQQDEwJjYTEhMB8GCSqGSIb3DQEJARYS
b3BlbnNzbEBodWF3ZWkuY29tMB4XDTA5MDQyNDExNTU0MFoXDTI5MDQxOTExNTU0
MFoweTELMAkGA1UEBhMCaW4xDDAKBgNVBAgTA2thcjEMMAoGA1UEBxMDYmFuMQ8w
DQYDVQQKEwZodWF3ZWkxDTALBgNVBAsTBGlwc2kxCzAJBgNVBAMTAmVlMSEwHwYJ
KoZIhvcNAQkBFhJvcGVuc3NsQGh1YXdlaS5jb20wgZ8wDQYJKoZIhvcNAQEBBQAD
gY0AMIGJAoGBAPrvtUQVDxt+0iYckFopQ6APWyqQXtgU7JZTHUET5mr0uu7QsnRx
g3HVf5/bQKl+sHTnagZYx+DuYJYfVXMGDyJqcVY1cGo/LxVznrVnRsPr2B+5QUNz
mz5X5eL2YARNCnzu1kaizKGC0PJDsPCsHNqrQl6AjrovoL1JU5yedTf3AgMBAAGj
ggGBMIIBfTAJBgNVHRMEAjAAMA4GA1UdDwEB/wQEAwIF4DBCBgNVHR8EOzA5MBqg
GKAWhhRodHRwOi8vbXkuY29tL215LmNybDAboBmgF4YVaHR0cDovL290aC5jb20v
bXkuY3JsMCwGCWCGSAGG+EIBDQQfFh1PcGVuU1NMIEdlbmVyYXRlZCBDZXJ0aWZp
Y2F0ZTAdBgNVHQ4EFgQUY/zr/dgtmQXs0gxUbz+mXLP0by0wga0GA1UdIwSBpTCB
ooAUAt0uKqoTKzstWrKlewmG48bdf6ihf6R9MHsxCzAJBgNVBAYTAmluMQwwCgYD
VQQIEwNrYXIxDDAKBgNVBAcTA2JhbjEPMA0GA1UEChMGaHVhd2VpMQ0wCwYDVQQL
EwRpcHNpMQ0wCwYDVQQDEwRyb290MSEwHwYJKoZIhvcNAQkBFhJvcGVuc3NsQGh1
YXdlaS5jb22CCQCsTg/o9PppkTAfBgNVHREEGDAWhwTAqAcBgg53d3cuaHVhd2Vp
LmNvbTANBgkqhkiG9w0BAQUFAAOBgQAqaKIREKiSmFR+HsxYKWiIP0oV0XEkRPIL
hwnyWTvjw1fkzjMIKvdmsA5z3PcBk2xNkqT2kaz+/DKTM1bAoVU2pV3SyEUylxdw
i4BZpnRRN1PWMfpuLxGbEo6gBJCMwPkjXDz8FlyTQBkoTq0iyl/Bfin6M2CL52Ca
+EhUNKUMpw==
-----END CERTIFICATE-----
subject=C = in, ST = kar, L = ban, O = huawei, OU = ipsi, CN = ee, emailAddress = openssl@xx.com

issuer=C = in, ST = kar, L = ban, O = huawei, OU = ipsi, CN = ca, emailAddress = openssl@xx.com

---
No client certificate CA names sent
---
SSL handshake has read 1108 bytes and written 270 bytes
Verification error: unable to get issuer certificate
---
New, SSLv3, Cipher is RC4-SHA
Server public key is 1024 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1
    Cipher    : RC4-SHA
    Session-ID:
    Session-ID-ctx:
    Master-Key: 63E6635F3FCA745441E375E9B9DB943F4532EED75727AE29D3A4A6C7315925A4B9CA84FC03B4C888376207CAC2CC54EE
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1605002284
    Timeout   : 7200 (sec)
    Verify return code: 2 (unable to get issuer certificate)
    Extended master secret: yes
---
`

**Comment by 1185129439** [comment]
![image](https://user-images.githubusercontent.com/69578444/98662660-13836c00-2383-11eb-92d8-f14396758667.png)
I have loaded the legacy provider,and set security level 0

**Comment by 1185129439** [comment]
Certificate is OK， I already verified by SSL3_CK_RSA_DES_192_CBC3_SHA and TLS1_CK_RSA_WITH_AES_256_SHA. Only RC4-SHA is  unsuccessful（bad record mac）

**Comment by mattcaswell** [comment]
This appears to be a bug. I can replicate this with latest master. Note, your comments above suggested you modified s_server/s_client to do this test. I was able to replicate it without any source changes as follows:

````
$ openssl s_server -tls1 -cipher "RC4-SHA:@SECLEVEL=0" -cert cert.pem -key key.pem -provider legacy -provider default
$ openssl s_client -tls1 -cipher "RC4-SHA:@SECLEVEL=0" -provider legacy -provider default
````
This causes s_server to fail as follows:
````
Using default temp DH parameters
ACCEPT
ERROR
8052FBAAE87F0000:error::SSL routines::decryption failed or bad record mac:ssl/record/ssl3_record.c:663:
shutting down SSL
CONNECTION CLOSED
````
Note, you need to compile with "enable-weak-ssl-ciphers" to be able to do this.

**Comment by slontis** [comment]
Seeing this issue here also..

**Comment by wbeck10** [comment]
I was investigating this issue a little.
The Mac is getting compared against a NULL buffer.
It seems the mac buffer wasn't set in cipher_generic_block_update
but tried to get read in tls1_enc for further comparing checks in ssl3_get_record:


It can be that the tlsversion is  set to late (set by ssl3_do_change_cipher_spec) but
cipher_generic_block_update only sets the Mac buffer if tlsversion > 0.

![Screen Shot 2020-11-11 at 11 01 08 am](https://user-images.githubusercontent.com/8453763/98752167-5853e500-240d-11eb-8d0d-843f5bee4c63.png)

**Comment by 1185129439** [comment]
> Seeing this issue here also..
sorry，I can not find the issue link

**Comment by slontis** [comment]
> > Seeing this issue here also..
> > sorry，I can not find the issue link

See @wbeck10 comment above.

**Comment by 1185129439** [comment]
> Note, you need to compile with "enable-weak-ssl-ciphers" to be able to do this.
I have compiled with “./config enable-weak-ssl-ciphers -d”

**Comment by 1185129439** [comment]
> I was investigating this issue a little.
> The Mac is getting compared against a NULL buffer.
> It seems the mac buffer wasn't set in cipher_generic_block_update
> but tried to get read in tls1_enc for further comparing checks in ssl3_get_record:
> 
> It can be that the tlsversion is set to late (set by ssl3_do_change_cipher_spec) but
> cipher_generic_block_update only sets the Mac buffer if tlsversion > 0.
> 
> ![Screen Shot 2020-11-11 at 11 01 08 am](https://user-images.githubusercontent.com/8453763/98752167-5853e500-240d-11eb-8d0d-843f5bee4c63.png)

What causes this result？and How to solve。

**Comment by 1185129439** [comment]
> This appears to be a bug. I can replicate this with latest master. Note, your comments above suggested you modified s_server/s_client to do this test. I was able to replicate it without any source changes as follows:
> 
> ```
> $ openssl s_server -tls1 -cipher "RC4-SHA:@SECLEVEL=0" -cert cert.pem -key key.pem -provider legacy -provider default
> $ openssl s_client -tls1 -cipher "RC4-SHA:@SECLEVEL=0" -provider legacy -provider default
> ```
> 
> This causes s_server to fail as follows:
> 
> ```
> Using default temp DH parameters
> ACCEPT
> ERROR
> 8052FBAAE87F0000:error::SSL routines::decryption failed or bad record mac:ssl/record/ssl3_record.c:663:
> shutting down SSL
> CONNECTION CLOSED
> ```
> 
> Note, you need to compile with "enable-weak-ssl-ciphers" to be able to do this.

I have compiled with “./config enable-weak-ssl-ciphers -d”

**Comment by 1185129439** [comment]
> > > Seeing this issue here also..
> > > sorry，I can not find the issue link
> 
> See @wbeck10 comment above.
Thanks. I think @wbeck10 just shows where the problem, but does not explain the causes of problems, also does not say the solution.

**Comment by 1185129439** [comment]
> > > Seeing this issue here also..
> > > sorry，I can not find the issue link
> 
> See @wbeck10 comment above.

Thanks. I think @wbeck10 just shows where the problem, but does not explain the causes of problems, also does not say the solution.

**Comment by mattcaswell** [comment]
I have a fix for this issue.. @wbeck10 is along the right lines, but RC4 is a stream cipher so the fix lies in `ossl_cipher_generic_stream_update` not in `ossl_cipher_generic_block_update`.

This bug was introduced because commit be9d82bb3 inadvertently broke the test that would have caught this. Fixing the test reveals another break in RC4-MD5 ciphers - which requires a different fix. I'm still working on that one, so no PR just yet.

**Comment by 1185129439** [comment]
> I have a fix for this issue.. @wbeck10 is along the right lines, but RC4 is a stream cipher so the fix lies in `ossl_cipher_generic_stream_update` not in `ossl_cipher_generic_block_update`.
> 
> This bug was introduced because commit [be9d82b](https://github.com/openssl/openssl/commit/be9d82bb35812ac65cd92316d1ae7c7c75efe9cf) inadvertently broke the test that would have caught this. Fixing the test reveals another break in RC4-MD5 ciphers - which requires a different fix. I'm still working on that one, so no PR just yet.

Thanks！When can you fix the bug.

**Comment by mattcaswell** [comment]
I'm hoping to work on it today. As a taster here is the critical change in master for the RC4-SHA issue. There are likely to be more changes somewhere around this code to fix the RC4-MD5 problems:

````  diff
diff --git a/providers/implementations/ciphers/ciphercommon.c b/providers/implementations/ciphers/ciphercommon.c
index 8d45d7a7d7..23f191fbbf 100644
--- a/providers/implementations/ciphers/ciphercommon.c
+++ b/providers/implementations/ciphers/ciphercommon.c
@@ -429,16 +429,27 @@ int ossl_cipher_generic_stream_update(void *vctx, unsigned char *out,
     }
 
     *outl = inl;
-    /*
-     * Remove any TLS padding. Only used by cipher_aes_cbc_hmac_sha1_hw.c and
-     * cipher_aes_cbc_hmac_sha256_hw.c
-     */
-    if (!ctx->enc && ctx->removetlspad > 0) {
-        /* The actual padding length */
-        *outl -= out[inl - 1] + 1;
+    if (!ctx->enc) {
+        /*
+        * Remove any TLS padding. Only used by cipher_aes_cbc_hmac_sha1_hw.c and
+        * cipher_aes_cbc_hmac_sha256_hw.c
+        */
+        if (ctx->removetlspad > 0) {
+            /* The actual padding length */
+            *outl -= out[inl - 1] + 1;
+
+            /* MAC and explicit IV */
+            *outl -= ctx->removetlspad;
+        }
 
-        /* MAC and explicit IV */
-        *outl -= ctx->removetlspad;
+        /* Extract the MAC if there is one */
+            /* MAC and explicit IV */
+            *outl -= ctx->removetlspad;
+        }
 
-        /* MAC and explicit IV */
-        *outl -= ctx->removetlspad;
+        /* Extract the MAC if there is one */
+        if (ctx->tlsmacsize > 0) {
+            if (*outl < ctx->tlsmacsize)
+                return 0;
+
+            ctx->tlsmac = out + *outl - ctx->tlsmacsize;
+            *outl -= ctx->tlsmacsize;
+        }
     }
 
     return 1;

````

**Comment by mattcaswell** [comment]
Proper fix for this in #13378.

**Comment by 1185129439** [comment]
> Proper fix for this in #13378.
thanks， I changed the code by alpha8 and used the same operation， but  a new problem appeared：
![image](https://user-images.githubusercontent.com/69578444/99027375-06939200-25a8-11eb-85c4-b47f86d59150.png)

**Comment by 1185129439** [comment]
> Proper fix for this in #13378.

Thanks！I changed the openssl3.0-alpha8 code by [https://github.com/openssl/openssl/pull/13378](url) , and I have checked many times code,  but here a new error. 
![image](https://user-images.githubusercontent.com/69578444/99043640-7bc38f00-25c9-11eb-8cf5-0495df169332.png)

**Comment by mattcaswell** [comment]
This isn't a bug. It means that OpenSSL tried to form the certificate chain but a certificate is missing. Check the "-CAfile" and "-CApath" options to OpenSSL - and check that you have all the correct certificates to form the chain.
